pkg/controller/staticpod/staticpod_controller.go

```diff
@@ -0,0 +1,671 @@
+//+kubebuilder:rbac:groups=apps.openyurt.io,resources=staticpods/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=apps.openyurt.io,resources=staticpods/finalizers,verbs=update
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile reads that state of the cluster for a StaticPod object and makes changes based on the state read
+// and what is in the StaticPod.Spec
+func (r *ReconcileStaticPod) Reconcile(_ context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	// Note !!!!!!!!!!
+	// We strongly recommend use Format() to  encapsulation because Format() can print logs by module
+	// @kadisi
+	klog.V(4).Infof(Format("Reconcile StaticPod %s", request.Name))
+
+	// Fetch the StaticPod instance
+	instance := &appsv1alpha1.StaticPod{}
+	if err := r.Get(context.TODO(), request.NamespacedName, instance); err != nil {
+		klog.Errorf("Fail to get StaticPod %v, %v", request.NamespacedName.Name, err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if instance.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	var (
+		// totalNumber represents the total number of nodes running the target static pod
+		totalNumber int32
+
+		// desiredNumber represents the desired upgraded number of nodes running the target static pod
+		// In auto mode: it's the number of ready nodes running the target static pod
+		// In ota mode: it's equal to totalNumber
+		desiredNumber int32
+
+		// upgradedNumber represents the number of nodes that have been upgraded
+		upgradedNumber int32
+	)
+
+	// The later upgrade operation is conducted based on upgradeInfos
+	upgradeInfos, err := upgradeinfo.New(r.Client, instance, UpgradeWorkerPodPrefix)
+	if err != nil {
+		klog.Errorf(Format("Fail to get static pod and worker pod upgrade info for nodes of StaticPod %v, %v",
+			request.NamespacedName.Name, err))
+		return ctrl.Result{}, err
+	}
+	totalNumber = int32(len(upgradeInfos))
+	// There are no nodes running target static pods in the cluster
+	if totalNumber == 0 {
+		klog.Infof(Format("No static pods need to be upgraded of StaticPod %v", request.NamespacedName.Name))
+		return r.updateStaticPodStatus(instance, totalNumber, totalNumber, totalNumber, upgradeSuccessCondition)
+	}
+
+	// The latest hash value for static pod spec
+	// This hash value is used in three places
+	// 1. Automatically added to the annotation of static pods to facilitate checking if the running static pods are up-to-date
+	// 2. Automatically added to the annotation of worker pods to facilitate checking if the worker pods are up-to-date
+	// 3. Added to static pods' corresponding configmap to facilitate checking if the configmap is up-to-date
+	latestHash := util.ComputeHash(&instance.Spec.Template)
```

Comment:
how about use the hash value for static pod annotations["kubernetes.io/config.hash"]?

Reply:
In staticpod controller, hash value is used to uniquely identify the manifest file. In kubelet, `kubernetes.io/config.hash` plays a similar role. However, `kubernetes.io/config.hash` is closely related to the node, which means that every edge static pod has a unique hash value.

Ref: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/config/common.go#L63

Example: (pods `nginx-kind-worker` and `nginx-kind-worker2` are using the same manifest)
![image](https://user-images.githubusercontent.com/60642406/229212916-f2d01890-efb0-4dd0-af92-28c15b9d6aaa.png)

If we use `kubernetes.io/config.hash` to reconcile, we need to calculate the hash value for every static pod which makes the control logic more complex. Especially we still need to determine a hash value to check the up-to-date status of the configmap.

Therefore, I suggest using the current practice(`openyurt.io/static-pod-hash`) rather than `kubernetes.io/config.hash`.

Reply:
@xavier-hou Thanks for your explanation, sorry i have not dived into the details of `kubernetes.io/config.hash`, and i agree with the idea of using `openyurt.io/static-pod-hash`.